### PR TITLE
Move statement errors out of the parser.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This PR specifically makes it so that rules about embedded statements and warnings about empty statements are not produced while parsing). Previously, we did this work at parse time. There's no reason for this complexity and specificity to be at the parser level. Syntactically we know what's going on, and we can allow this while parsing and check and report the error later. It also prevents incrementally reusing statements (and their containers) when this happens.

Tagging @dotnet/roslyn-compiler @gafter @VSadov

This part of a long term effort to remove unnecessary errors from the parser and move them to later stages of the compiler pipeline. Doing this makes the parser simpler, makes it faster to get a syntax tree, and makes our incremental implementation simpler (because the parser/incremental-parser needs to track less contextual states).

This is also important as syntax-diagnostics end up causing issues for higher level IDE features.

**Comment by CyrusNajmabadi** [comment]
Tagging @TyOverby and @jcouv as well.  Thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -7600,41 +7602,9 @@ private bool IsEndOfFixedStatement()
...
-            {
-                statement = this.ParseStatementCore();
-                statement = this.AddError(statement, ErrorCode.WRN_PossibleMistakenNullStatement);
-            }
-            else
-            {
-                statement = this.ParseStatementCore();
-            }
-
-            // The consumers of embedded statements are expecting to receive a non-null statement yet there are
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >// The consumers of embedded statements are expecting to receive a non-null statement yet there are  [](start = 12, length = 100)

Please preserve this comment.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -8156,37 +8128,33 @@ private GotoStatementSyntax ParseGotoStatement()
...
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.IfKeyword);
-            var @if = this.EatToken(SyntaxKind.IfKeyword);
-            var openParen = this.EatToken(SyntaxKind.OpenParenToken);
-            var condition = this.ParseExpressionCore();
-            var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-            var statement = this.ParseEmbeddedStatement(false);
-            var elseClause = ParseElseClauseOpt();
-
-            return _syntaxFactory.IfStatement(@if, openParen, condition, closeParen, statement, elseClause);
+            return _syntaxFactory.IfStatement(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >return _syntaxFactory.IfStatement( [](start = 12, length = 34)

Please restore intermediate locals, the code is easier to debug that way.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -7974,10 +7946,10 @@ private ForStatementSyntax ParseForStatement()
                     this.ParseForStatementExpressionList(ref semi2, incrementors);
                 }
 
-                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-                var statement = ParseEmbeddedStatement(true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var statement = ParseEmbeddedStatement(true); [](start = 16, length = 45)

Please restore intermediate locals, the code is easier to debug that way.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -8156,37 +8128,33 @@ private GotoStatementSyntax ParseGotoStatement()
...
 
         private ElseClauseSyntax ParseElseClauseOpt()
         {
-            if (this.CurrentToken.Kind != SyntaxKind.ElseKeyword)
-            {
-                return null;
-            }
-
-            var elseToken = this.EatToken(SyntaxKind.ElseKeyword);
-            var elseStatement = this.ParseEmbeddedStatement(false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var elseStatement = this.ParseEmbeddedStatement(false); [](start = 12, length = 55)

Please restore intermediate locals, the code is easier to debug that way.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -8156,37 +8128,33 @@ private GotoStatementSyntax ParseGotoStatement()
...
         }
 
         private LockStatementSyntax ParseLockStatement()
         {
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.LockKeyword);
-            var @lock = this.EatToken(SyntaxKind.LockKeyword);
-            var openParen = this.EatToken(SyntaxKind.OpenParenToken);
-            var expression = this.ParseExpressionCore();
-            var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-            var statement = this.ParseEmbeddedStatement(false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var statement = this.ParseEmbeddedStatement(false); [](start = 12, length = 51)

Please restore intermediate locals, the code is easier to debug that way.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -8383,10 +8351,10 @@ private UsingStatementSyntax ParseUsingStatement()
             ParseUsingExpression(ref declaration, ref expression, ref resetPoint);
             this.Release(ref resetPoint);
 
-            var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-            var statement = this.ParseEmbeddedStatement(false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var statement = this.ParseEmbeddedStatement(false); [](start = 12, length = 51)

Please restore intermediate locals, the code is easier to debug that way.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -8486,12 +8454,12 @@ private bool IsUsingStatementVariableDeclaration(ScanTypeFlags st)
         private WhileStatementSyntax ParseWhileStatement()
         {
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.WhileKeyword);
-            var @while = this.EatToken(SyntaxKind.WhileKeyword);
-            var openParen = this.EatToken(SyntaxKind.OpenParenToken);
-            var condition = this.ParseExpressionCore();
-            var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-            var statement = this.ParseEmbeddedStatement(true);
-            return _syntaxFactory.WhileStatement(@while, openParen, condition, closeParen, statement);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >return _syntaxFactory.WhileStatement(@while, openParen, condition, closeParen, statement); [](start = 12, length = 90)

Please restore intermediate locals, the code is easier to debug that way.

**Comment by CyrusNajmabadi** [comment]
What is the policy on locals vs nested calls?  I see all sorts of patterns here in the parser for how this type of code is written.  It doesn't look like there's any set rule to provide a local variable for all intermediately generated values.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 8]
```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
...
+
+                    case SyntaxKind.ForStatement:
+                    case SyntaxKind.ForEachStatement:
+                    case SyntaxKind.WhileStatement:
+                        if (node.SemicolonToken.GetNextToken().Kind() == SyntaxKind.OpenBraceToken)
+                        {
+                            // For loop constructs, only warn if we see a block following the statement.
+                            // That indicates code like:  "while (x) ; { }"
+                            // which is most likely a bug.
+                            diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation()); [](start = 28, length = 81)

goto default?

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 9]
```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
...
+            return false;
+        }
+
+        private static BoundStatement BindEmpty(EmptyStatementSyntax node, DiagnosticBag diagnostics)
+        {
+            if (IsEmbeddedStatementParent(node.Parent))
+            {
+                switch (node.Parent.Kind())
+                {
+                    default:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: > default: [](start = 19, length = 9)

Consider keeping default case last.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 10]
```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
             return new BoundThrowStatement(node, boundExpr, hasErrors);
         }
 
-        private static BoundStatement BindEmpty(EmptyStatementSyntax node)
+        private static bool IsEmbeddedStatementParent(SyntaxNode node)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >IsEmbeddedStatementParent [](start = 28, length = 25)

It feels like we don't need a helper like this, it will be sufficient to perform necessary checks in BindPossibleEmbeddedStatement. In there we can be sure that statements of the interest (an empty statement, declarations, labeled statement) are embedded statements. At the same time, no embedded statement should bypass that method.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 11]
```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
...
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static BoundStatement BindEmpty(EmptyStatementSyntax node, DiagnosticBag diagnostics)
+        {
+            if (IsEmbeddedStatementParent(node.Parent))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >if (IsEmbeddedStatementParent(node.Parent)) [](start = 12, length = 43)

It feels like it would be better to intercept this statement in BindPossibleEmbeddedStatement function instead.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 12]
```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
...
+                        break;
+                }
+            }
+
             return new BoundNoOpStatement(node, NoOpStatementFlavor.Default);
         }
 
         private BoundLabeledStatement BindLabeled(LabeledStatementSyntax node, DiagnosticBag diagnostics)
         {
+            if (IsEmbeddedStatementParent(node.Parent))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >if (IsEmbeddedStatementParent(node.Parent)) [](start = 12, length = 43)

It feels like it would be better to intercept this statement in BindPossibleEmbeddedStatement function instead.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 13]
```diff
@@ -442,6 +494,12 @@ private BoundStatement BindGoto(GotoStatementSyntax node, DiagnosticBag diagnost
 
         private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax node, DiagnosticBag diagnostics)
         {
+            if (IsEmbeddedStatementParent(node.Parent))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >if (IsEmbeddedStatementParent(node.Parent)) [](start = 12, length = 43)

It feels like it would be better to intercept this statement in BindPossibleEmbeddedStatement function instead.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 14]
```diff
@@ -542,6 +600,12 @@ private void CheckForUnobservedAwaitable(BoundExpression expression, DiagnosticB
 
         internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
         {
+            if (IsEmbeddedStatementParent(node.Parent))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >if (IsEmbeddedStatementParent(node.Parent)) [](start = 12, length = 43)

It feels like it would be better to intercept this statement in BindPossibleEmbeddedStatement function instead.

**Comment by AlekseyTs** [comment]
> What is the policy on locals vs nested calls?

The policy is very simple: 
We have very small number of rules that should be followed. Outside of that, when you write new code, you are free to choose. When making changes to an existing code, do not change the style just because you think your style is better. I already asked you before to not make style changes to the existing code. I will be asking you undo changes like that.

**Comment by AlekseyTs** [comment]
>         public void TestRegressPragmaWarningDisableWithWarningCode()

It is not clear why this test must be modified. Could you explain please?

---
Refers to: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs:3327 in a4614ca. [](commit_id = a4614ca0ea89e66b5c85dfebbdc8f87ff8808cb6, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 12 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 13 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 14 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Done

**Comment by jaredpar** [comment]
> actually adopt styles, and use that style across all the code.

As for coding style the Roslyn code base as a whole (compiler, IDE, EE, etc ...) follows the dotnet coding guidelines.  

- https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

This was agreed to when the code base when OSS.  What style we follow isn't worth debating.  The style is documented there and should be followed when possible.  

> if a style is not adopted, then allow code to be changed, so that it becomes more consistent with other code.

Sure but mixing style + functionality change makes it more difficult for the team to review.  I'd personally prefer to see them as separate PRs.

**Comment by CyrusNajmabadi** [comment]
@jaredpar 

> Sure but mixing style + functionality change makes it more difficult for the team to review. I'd personally prefer to see them as separate PRs.

I'm personally not a fan of this because, IME, it means that no progress is ever made on hte consistency front.

I prefer to follow the model we've taken on the IDE side where people can (and do) adjust code consistency as they're working in the code.  With this approach you actually take those steps (small as they may be) that actually end up having a large positive impact on consistency.

It's as i mentioned above.  The approach taken actually actively works to subvert a desire to make the code consistent and actually follow styles the team has decided on.  We saw this with the old native compiler as well.  Practically zero effort was ever actually taken to address these sorts of issues, and so the codebase festered.  Eventually it gets to a point where anyone else trying to come in and work in the codebase faces enormous challenges even figuring out how they should be coding.

--

I personally believe this has actual implications for code quality.  As an example, it's practically impossible for anyone to come into the compiler and know the rules that one should follow since all the code is extremely inconsistent in how it was written.  The more you make a set of code consistent with itself, the more that improperly written code will stand out as incorrect.   Also, the more conssitent you make code, the easier it is to extract out common structure that enables the remainder of hte code to be written more simply.

This is not hypothetical either.  As i've mentioned before, the TypeScript parser is only 5k lines (including incremental). Roslyn's C# parser is nearly 8k lines, and doesn't even handle incremental in that count.  This line divergence exists because of hte massive amount of excessive inconsistency present in the parser and major amount of duplication that that then causes.  i.e. we should be able to share far more code.  But it's extremely difficult to even realize because the parser doesn't take care to use patterns and styles that make it clear what could be shared.  Even though the languages have similar complexity (TS is actually probably more complex).

--

This is one of the reasons i'm sending through so many PRs in this space.  I want to get Roslyn's parser to be as good as i know it can be.  But it's going to take a lot of steps.  And it's *much* simpler if i can clean up as i go along and run into the issues.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 15]
```diff
@@ -7768,10 +7743,10 @@ private ForStatementSyntax ParseForStatement()
                     this.ParseForStatementExpressionList(ref semi2, incrementors);
                 }
 
-                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-                var statement = ParseEmbeddedStatement(true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var statement = ParseEmbeddedStatement(true); [](start = 16, length = 45)

Please restore intermediate locals.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 16]
```diff
@@ -360,7 +397,7 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
             return new BoundThrowStatement(node, boundExpr, hasErrors);
         }
 
-        private static BoundStatement BindEmpty(EmptyStatementSyntax node)
+        private static BoundStatement BindEmpty(EmptyStatementSyntax node, DiagnosticBag diagnostics)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >, DiagnosticBag diagnostics [](start = 73, length = 27)

Consider removing added parameter, which is not used.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Done!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 15 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: #fixed

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: #fixed

**Comment by CyrusNajmabadi** [comment]
>It is not clear why this test must be modified. Could you explain please?

Test needed to be modified because it was specifically seeing if a parser warning would be disabled when you had a pragma.  The specific warning this tested was moved to the binder.  As such, the test of the parser API was no longer relevant.  I updated the code so that we would continue testing an actual warning still produced by the parser.

**Comment by CyrusNajmabadi** [comment]
This PR is ready for review again @AlekseyTs

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 17]
```diff
@@ -7331,38 +7333,15 @@ private bool IsEndOfFixedStatement()
...
-            {
-                Debug.Assert(CurrentToken.Kind != SyntaxKind.SemicolonToken);
-                statement = SyntaxFactory.EmptyStatement(EatToken(SyntaxKind.SemicolonToken));
-            }
+            // The consumers of embedded statements are expecting to receive a non-null statement 
+            // yet there are several error conditions that can lead ParseStatementCore to return 
+            // null.  When that occurs create an error empty Statement and return it to the caller.
+            StatementSyntax statement = this.ParseStatementCore() ?? SyntaxFactory.EmptyStatement(EatToken(SyntaxKind.SemicolonToken));
 
             switch (statement.Kind)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Consider using `if` instead of `switch` since single case.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ForEachTests.cs [hunk 18]
```diff
@@ -1900,8 +1900,7 @@ public static void Main(string [] args)
                 Diagnostic(ErrorCode.ERR_IllegalStatement, "i < 5").WithLocation(6, 18),
                 // (6,25): error CS0103: The name 'i' does not exist in the current context
                 //     foreach(int; i < 5; i++)
-                Diagnostic(ErrorCode.ERR_NameNotInContext, "i").WithArguments("i").WithLocation(6, 25)
-                );
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "i").WithArguments("i").WithLocation(6, 25));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ForEachTests.cs]: Consider reverting since only change in file. The old format is actually more convenient when diagnostics need to be updated.

**Comment by CyrusNajmabadi** [comment]
Rebasing to 15.6

**Comment by gafter** [comment]
@CyrusNajmabadi Note that you have two sign-offs from the compiler team.

**Comment by CyrusNajmabadi** [comment]
I wasn't certain if @AlekseyTs wanted to take another look.    @AlekseyTs you mentioned #closed before.  Does that mean you're done with the review?

(note: please put this after anything you need to do for 15.3, this is not critical work).

**Comment by gafter** [comment]
#closed is how CodeFlow tags a comment when you mark it closed.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs [hunk 19]
```diff
@@ -3425,18 +3425,18 @@ public void TestRegressPragmaWarningDisableWithWarningCode()
             var text = @"
 class A
 {
-    static void Main(bool b)
+    static void Main(int i)
     {
-#pragma warning disable 642
-        if (b);{}
+#pragma warning disable 1522
+        switch (i) { }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs]: >switch (i) { } [](start = 8, length = 14)

I am not sure what is the motivation behind the change in this test.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 20]
```diff
@@ -296,6 +305,34 @@ internal BoundStatement BindPossibleEmbeddedStatement(StatementSyntax node, Diag
...
+                            {
+                                break;
+                            }
+
+                            goto default;
+
+                        default:
+                            // For non-loop constructs, always warn.  This is for code like:
+                            // "if (x) ;" which is almost certainly a bug.
+                            diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation()); [](start = 28, length = 81)

It feels like this error should be reported only one the semicolon token is present in code, i.e. is not missing, but it doesn't look like we are making this check. Could you please add a test for a scenario like this?

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs (hunk 19 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs]: Should we test both scenarios?

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs (hunk 19 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/PreprocessorTests.cs]: Sorry, shoudl have explained.  This is a test that is validating that ```#pragma warning disable``` works at the syntax layer.  However, "if (b); {}" moved from being a warning produced in the syntax layer to one produced in the semantic layer.  As such the test failed.  As the test is about pragma warning disabling, and not about htat particular warning, i changed the test to another syntactic form that still reports the warning during parsing.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 20 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Yup.  Will do.

**Comment by CyrusNajmabadi** [comment]
Missing-semicolon case checked for, and test added.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 21]
```diff
@@ -5155,6 +5155,37 @@ public static int Main(System.Collections.IEnumerable e)
...
+class MyClass
+{
+    public static int Main(bool b)
+    {
+        if (b)
+    
+    public
+";
+
+            CreateStandardCompilation(test).VerifyDiagnostics(
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: validated that without the codechange that this test would report the warning.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: PR is ready for rereview.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
(retesting as last commit was 18 days ago).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 22]
```diff
@@ -7876,21 +7855,21 @@ private IfStatementSyntax ParseIfStatement()
...
+            var statement = this.ParseEmbeddedStatement();
             var elseClause = ParseElseClauseOpt();
 
             return _syntaxFactory.IfStatement(@if, openParen, condition, closeParen, statement, elseClause);
         }
 
         private ElseClauseSyntax ParseElseClauseOpt()
         {
-            if (this.CurrentToken.Kind != SyntaxKind.ElseKeyword)
+           if (this.CurrentToken.Kind != SyntaxKind.ElseKeyword)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: wrong indent
